libs/langchain/langchain/vectorstores/tiledb.py

```diff
@@ -0,0 +1,771 @@
+"""Wrapper around TileDB vector database."""
+from __future__ import annotations
+
+import pickle
+import random
+import sys
+from typing import Any, Dict, Iterable, List, Mapping, Optional, Tuple
+
+import numpy as np
+
+from langchain.docstore.document import Document
+from langchain.schema.embeddings import Embeddings
+from langchain.schema.vectorstore import VectorStore
+from langchain.vectorstores.utils import maximal_marginal_relevance
+
+INDEX_METRICS = frozenset(["euclidean"])
+DEFAULT_METRIC = "euclidean"
+DOCUMENTS_ARRAY_NAME = "documents"
+VECTOR_INDEX_NAME = "vectors"
+MAX_UINT64 = np.iinfo(np.dtype("uint64")).max
+MAX_FLOAT_32 = np.finfo(np.dtype("float32")).max
+MAX_FLOAT = sys.float_info.max
+
+
+def dependable_tiledb_import() -> Any:
+    """Import tiledb-vector-search if available, otherwise raise error."""
+    try:
+        import tiledb as tiledb
+        import tiledb.vector_search as tiledb_vs
+    except ImportError:
+        raise ValueError(
+            "Could not import tiledb-vector-search python package. "
+            "Please install it with `conda install -c tiledb tiledb-vector-search` "
+            "or `pip install tiledb-vector-search`"
+        )
+    return tiledb_vs, tiledb
+
+
+def get_vector_index_uri(uri: str) -> str:
+    return f"{uri}/{VECTOR_INDEX_NAME}"
+
+
+def get_documents_array_uri(uri: str) -> str:
+    return f"{uri}/{DOCUMENTS_ARRAY_NAME}"
+
+
+class TileDB(VectorStore):
+    """Wrapper around TileDB vector database.
+
+    To use, you should have the ``tiledb-vector-search`` python package installed.
+
+    Example:
+        .. code-block:: python
+
+            from langchain import TileDB
+            embeddings = OpenAIEmbeddings()
+            db = TileDB(embeddings, index_uri, metric)
+
+    """
+
+    def __init__(
+        self,
+        embedding: Embeddings,
+        index_uri: str,
+        metric: str,
```

Comment:
nit: can we add `*` to signature after metric

Reply:
Applied.